README.md
# apiHandler
A dart library that connects to interact with the Ethereum blockchain. It connects
to an Ethereum node to send transactions, interact with smart contracts and much
more!

### Features
- Connect to an Ethereum node with the rpc-api, call common methods
- Send signed Ethereum transactions
- Generate private keys, setup new Ethereum addresses
- Call functions on smart contracts and listen for contract events
### TODO
- Code generation based on smart contract ABI for easier interaction
- Encode all supported solidity types, although only (u)fixed,
  which are not commonly used, are not supported at the moment.

## Usage

### Credentials and Wallets
In order to send transactions on the Ethereum network, some credentials
are required. The library supports raw private keys and v3 wallet files.

### Only support rinkeby testnet now

```dart
import 'package:transaction/web3dart.dart';
// You can setup private keys first
web3dart.initEthClient(privateKey:'d1bdc683fbeb9fa0b4ceb26adb39eaffb21b16891ea28e4cf1bc3118fdd39295');
```

## Sending transactions
Of course, this library supports creating, signing and sending Ethereum
transactions:

```dart
import 'package:web3dart/web3dart.dart';
/// send ETH Token Transaction
web3dart.sendETHFromMainnet(amount: EtherAmount.fromUnitAndValue(EtherUnit.finney, 1),toAddress:'0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A');

/// send ERC20 Token Transaction
/// if usdt Decimals is 18 than 1000000000000000000 = 1 amount  for example: BigInt.from(1000000000000000000)
web3dart.sendERC20FromMainnet(contractAddress: '0x7e0480ca9fd50eb7a3855cf53c347a1b4d6a2ff5',amount: BigInt.from(1000000000000000000),toAddress:'0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A');

```

## get TopErc20CoinPrice()

```dart
import 'package:transaction/model/coinPrice.dart';
import 'package:transaction/transaction.dart';
import 'dart:async';

/// getTopErc20CoinPrice
CoinPriceList coinPrice = await transaction.getTopErc20CoinPrice();
    print("result:$coinPrice");

```


CHANGELOG.md
## 0.4.9
- update flutter version to 2.2.3

## 0.4.8
- add transferRequest method
- add EnvType Dev,Staging,Testing

## 0.4.7
- fix some bug

## 0.4.6
- fix some bug

## 0.4.5
- fix some bug

## 0.4.4
- fix some bug

## 0.4.3
- update eurus to dev env

## 0.4.2
- back to use infura rinkeby rpc

## 0.4.1
- update rinkebyRpc url

## 0.4.0
- estimatenErc20Gas to more accurate
- low medium fast  transaction speed function implementation

## 0.3.8
- update fixed gas Price on testnet

## 0.3.7
- update gas Price 

## 0.3.6
- backend updated smart contract addresses of Rinkeby EurusInternalConfig and Eurus ExternalSmartContractConfig

## 0.3.5
- add canGetPrivateKeyHandler method

## 0.3.4
- add getTokenSymbol method

## 0.3.3
- add submitWithdrawERC20 method
- add getEurusUserDepositAddress method

## 0.3.2
- add get decimal method from smart contract

## 0.3.1
- add setErc20Contract

## 0.3.0
- update balance string

## 0.2.9

- add getBalance method.

## 0.2.8

- support estimateErcTokenGas method.

## 0.2.7

- only use blockChainType parameter to choose blockchain

## 0.2.6

- add getERC20Balance method
- add getETHBalance method

## 0.2.5

- update web3dart version to 2.0.0-dev.9


## 0.2.4

- change to rinkeby testnet

## 0.2.3

- show estimateGas on plugin on example

## 0.2.2

- update abi

## 0.2.1

- remove gas price on mainnet transaction

## 0.2.0

- remove not use code

## 0.1.9

- add transfer eurus usdt function

## 0.1.8

- fix some bug

## 0.1.7

- support getting current ethereum address function

## 0.1.6

- support eurus eth client and mainnet eth client

## 0.1.5

- support eurus blockchain eth transaction

## 0.1.4

- add return type into method

## 0.1.3

- remove gas price to use default gas price

## 0.1.2

- update getTopErc20CoinPrice function from coingecko.com
- add comment

## 0.1.1

- add getTopErc20CoinPrice function from coingecko.com

## 0.1.0

- Initial version, created by ken

example/lib/main.dart
// import 'package:transaction/model/coinPrice.dart';
// import 'package:transaction/transaction.dart';
import 'package:transaction/web3dart.dart';
import 'package:flutter/material.dart';
import 'dart:async';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int _counter = 0;

  @override
  void initState() {
    super.initState();
    web3dart.initEthClient(
        privateKey:
            'd1bdc683fbeb9fa0b4ceb26adb39eaffb21b16891ea28e4cf1bc3118fdd39295');
    // web3dart.getOwnerAddress();
  }

  Future<void> _incrementCounter() async {
    // web3dart.getETHClientDetail();
    // web3dart.getAddressDetail();
    print("ethereumAddress:${web3dart.myEthereumAddress.toString()}");
    //if usdt Decimals is 6 than 100000 = 1 amount   like BigInt.from(1000000)
    // web3dart.getERC20Balance(blockChainType: BlockChainType.Eurus,deployedContract: web3dart.usdtContractFromEurus,decimals: 6);
    // web3dart.getERC20Balance(blockChainType: BlockChainType.Ethereum,deployedContract: web3dart.usdtContractFromEthereum,decimals: 18);
    // web3dart.getETHBalance(blockChainType: BlockChainType.Eurus);
    // web3dart.getETHBalance(blockChainType: BlockChainType.Ethereum);
    // String estimateGas = await web3dart.estimateGas(toAddress: '0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A',amount: BigInt.from(100000000000000),blockChainType: BlockChainType.Ethereum);
    // print("estimateGas:$estimateGas + ETH");
    // web3dart.sendETH(amount: BigInt.from(10000000000000000),toAddress:'0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A',type: BlockChainType.Ethereum);
    if (web3dart.erc20ContractFromEthereum != null) {
      web3dart.estimateErcTokenGas(
          blockChainType: BlockChainType.Ethereum,
          deployedContract: web3dart.erc20ContractFromEthereum!,
          amount: BigInt.from(100000000000000),
          toAddress: '0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A');
      web3dart.sendERC20(
          deployedContract: web3dart.erc20ContractFromEthereum!,
          enterAmount: 1,
          toAddress: '0xA3B4dE5E90A18512BD82c1A640AC99b39ef2258A',
          blockChainType: BlockChainType.Ethereum);
    }
    // CoinPriceList coinPrice = await transaction.getTopErc20CoinPrice();
    // print("result:$coinPrice");
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text(
            '$_counter',
            style: Theme.of(context).textTheme.headline4,
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: Icon(Icons.add),
        ), // This trail
      ),
    );
  }
}
